File: org.opencompare/api-java/src/main/java/org/opencompare/api/java/io/ExportMatrixExporter.java
package org.opencompare.api.java.io;

import org.jsoup.nodes.Element;
import org.jsoup.parser.Tag;
import org.opencompare.api.java.*;
import org.opencompare.api.java.util.Pair;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by gbecan on 08/12/15.
 */
public class ExportMatrixExporter {

    public ExportMatrix export(PCMContainer pcmContainer) {
        PCM pcm = pcmContainer.getPcm();

        final PCMMetadata metadata;
        if (pcmContainer.getMetadata() == null) {
            metadata = new PCMMetadata(pcm);
        } else {
            metadata = pcmContainer.getMetadata();
        }


        ExportMatrix matrix = new ExportMatrix();

        // Export name
        matrix.setName(pcm.getName());

        int productsStartRow = exportFeatures(metadata, matrix);

        exportProducts(pcm, metadata, matrix, productsStartRow);

        // Transpose matrix if necessary
        if (!metadata.getProductAsLines()) {
            matrix.transpose();
        }

        return matrix;
    }


    private int exportFeatures(PCMMetadata metadata, ExportMatrix matrix) {
        List<Pair<AbstractFeature, Pair<Integer, Integer>>> currentFeatureLevel = new ArrayList<>();

        for (Feature feature : metadata.getSortedFeatures()) {
            currentFeatureLevel.add(new Pair<>(feature, new Pair<>(1, 1)));
        }

        List<List<ExportCell>> exportCellRows = new ArrayList<>();
        boolean noParents = false;

        while(!currentFeatureLevel.isEmpty() && !noParents) {

            List<Pair<AbstractFeature, Pair<Integer, Integer>>> nextFeatureLevel = new ArrayList<>();
            List<Pair<AbstractFeature, Pair<Integer, Integer>>> row = new ArrayList<>();

            // Detect if current level of features have at least one parent
            noParents = true;
            for (Pair<AbstractFeature, Pair<Integer, Integer>> data : currentFeatureLevel) {
                if (data._1.getParentGroup() != null) {
                    noParents = false;
                }
            }

            // Analyze hierarchy of features
            int i = 0;
            while (i < currentFeatureLevel.size()) {
                Pair<AbstractFeature, Pair<Integer, Integer>> data = currentFeatureLevel.get(i);
                AbstractFeature aFeature = data._1;
                Pair<Integer, Integer> span = data._2;

                // Compute colspan
                int colspan = 1;
                while ((i + 1) < currentFeatureLevel.size() && aFeature.equals(currentFeatureLevel.get(i + 1)._1)) {
                    i++;
                    colspan++;
                }

                // Compute rowspan and prepare for next iteration
                AbstractFeature parentGroup = aFeature.getParentGroup();
                if (parentGroup == null) {
                    int rowspan = span._1 + 1;
                    nextFeatureLevel.add(new Pair<>(aFeature, new Pair<>(rowspan, colspan)));
                    if (noParents) {
                        row.add(new Pair<>(aFeature, new Pair<>(span._1, colspan)));
                    }
                } else {
                    row.add(new Pair<>(aFeature, new Pair<>(span._1, colspan)));
                    nextFeatureLevel.add(new Pair<>(parentGroup, new Pair<>(1, 1)));
                }
                i++;
            }

            // Create cells
            List<ExportCell> exportCellRow = new ArrayList<>();
            for (Pair<AbstractFeature, Pair<Integer, Integer>> data : row) {
                AbstractFeature aFeature = data._1;
                Pair<Integer, Integer> span = data._2;

                ExportCell exportCell = new ExportCell(aFeature.getName(), span._1, span._2);
                exportCell.setFeature(true);
                exportCell.setInProductsKeyColumn(false);

                exportCellRow.add(exportCell);
            }
            exportCellRows.add(exportCellRow);

            currentFeatureLevel = nextFeatureLevel;
        }

        // Add rows to table
        Collections.reverse(exportCellRows);
        int row = 0;
        for (List<ExportCell> exportCellRow : exportCellRows) {

            int column = 0;

            // Skip cells created by rowspan
            while (matrix.isPositionOccupied(row, column)) {
                column++;
            }

            for (ExportCell exportCell : exportCellRow) {
                matrix.setCell(exportCell, row, column);
                column += exportCell.getColspan();
            }

            row++;
        }

        return row;
    }

    private void exportProducts(PCM pcm, PCMMetadata metadata, ExportMatrix matrix, int productsStartRow) {
        int row = productsStartRow;
        for (Product product : metadata.getSortedProducts()) {

            int column = 0;
            for (Feature feature : metadata.getSortedFeatures()) {
                Cell cell = product.findCell(feature);

                if (cell != null) {
                    ExportCell exportCell = new ExportCell(cell.getContent(), cell.getRawContent());
                    exportCell.setFeature(false);
                    exportCell.setInProductsKeyColumn(pcm.getProductsKey().equals(feature));
                    matrix.setCell(exportCell, row, column);
                }

                column++;
            }

            row++;
        }
    }
}

File: org.opencompare/api-scala/src/main/scala/org/opencompare/api/scala/io/ExportMatrixExporter.scala
package org.opencompare.api.scala.io

import org.opencompare.api.scala.{AbstractFeature, PCM}
import org.opencompare.api.scala.metadata.{Orientation, Positions, ProductsAsColumns}

class ExportMatrixExporter {

  def export(pcm: PCM with Positions with Orientation) : ExportMatrix = {

    val matrix = new ExportMatrix

    matrix.name = pcm.name

    val productsStartRow = exportFeatures(pcm, matrix)

    exportProducts(pcm, matrix, productsStartRow)

    // Transpose matrix if necessary
    pcm.orientation match {
      case ProductsAsColumns() => matrix.transpose()
      case _ =>
    }
    matrix
  }

  def exportFeatures(pcm : PCM with Positions, matrix : ExportMatrix) : Int = {

    var currentFeatureLevel = List.empty[(AbstractFeature, Int, Int)]

    for (feature <- pcm.sortedFeatures()) {
      currentFeatureLevel = currentFeatureLevel :+ (feature, 1, 1)
    }

    var exportCellRows = List.empty[List[ExportCell]]
    var noParents = false

    while (currentFeatureLevel.nonEmpty && !noParents) {
      var nextFeatureLevel = List.empty[(AbstractFeature, Int, Int)]
      var row = List.empty[(AbstractFeature, Int, Int)]

      // Detect if current level of features has at least one parent
      noParents = currentFeatureLevel.forall(l => l._1.parent.isEmpty)

      // Analyze hierarchy of features
      var i = 0
      while (i < currentFeatureLevel.size) {
        val (feature, rowspan, colspan) = currentFeatureLevel(i)

        // Compute colspan
        var newColspan = 1
        while ((i + 1) < currentFeatureLevel.size && (feature == currentFeatureLevel(i + 1)._1)) {
          i += 1
          newColspan += 1
        }

        // Compute rowspan and prepare for next iteration
        val parentGroup = feature.parent

        if (parentGroup.isEmpty) {
          val newRowspan = rowspan + 1
          nextFeatureLevel = nextFeatureLevel :+ (feature, newRowspan, newColspan)
          if (noParents) {
            row = row :+ (feature, rowspan, newColspan)
          }
        } else {
          row = row :+ (feature, rowspan, newColspan)
          nextFeatureLevel = nextFeatureLevel :+ (parentGroup.get, 1, 1)
        }

        i += 1
      }

      // Create cells
      val exportCellRow = for ((feature, rowspan, colspan) <- row) yield {
        val exportCell = new ExportCell(feature.name, feature.name, rowspan, colspan)
        exportCell.feature = true
        exportCell.inProductsKeyColumn = false
        exportCell
      }

      exportCellRows = exportCellRows :+ exportCellRow

      currentFeatureLevel = nextFeatureLevel
    }

    // Add rows to table
    for ((exportCellRow, row) <- exportCellRows.reverse.zipWithIndex) {
      var column = 0
      while (matrix.isPositionOccupied(row, column)) {
        column += 1
      }

      for (exportCell <- exportCellRow) {
        matrix.setCell(exportCell, row, column)
        column += exportCell.colspan
      }
    }

    exportCellRows.size
  }

  def exportProducts(pcm : PCM with Positions, matrix : ExportMatrix, productsStartRow : Int): Unit = {
    var row = productsStartRow
    for (product <- pcm.sortedProducts()) {
      var column = 0

      for (feature <- pcm.sortedFeatures()) {
        val cellOpt = product.findCell(feature)

        cellOpt.foreach { cell =>
          val exportCell = new ExportCell(cell.content, cell.rawContent)
          exportCell.feature = false
          exportCell.inProductsKeyColumn = pcm.productsKey == feature
          matrix.setCell(exportCell, row, column)
        }

        column += 1
      }

      row += 1
    }
  }

}
